fml/raster_thread_merger.cc

```diff
@@ -64,12 +78,16 @@ void RasterThreadMerger::MergeWithLease(size_t lease_term) {
     return;
   }
 
-  bool success = shared_merger_->MergeWithLease(this, lease_term);
-  if (success && merge_unmerge_callback_ != nullptr) {
-    merge_unmerge_callback_();
-  }
+  // the critical section of thead merging
+  {
+    ScopedCriticalSection lock(critical_section_.get());
```

Comment:
Rather than this, could we just flush the task runners to make sure nothing else has added itself behind our back? Am I missing something here?

Reply:
The platform thread needs to wait for the raster thread to complete some tasks synchronously, for example, `platform_view->CreateRenderingSurface`. During the synchronous call, if the two threads are merged, this means the task queue of the raster thread also needs to be processed by the platform thread. But, the platform thread is already locked by itself and never gets a task from the task queue of the raster thread.

So, I added a synchronization mechanism here to avoid deadlock.

Reply:
For example:

```c++
void TaskRunner::RunNowOrPostTask(fml::RefPtr<fml::TaskRunner> runner,
                                  const fml::closure& task) {
  FML_DCHECK(runner);
  if (runner->RunsTasksOnCurrentThread()) { /// -------->>> A
    task();
  } else {
    runner->PostTask(std::move(task));   /// ----------->> B
  }
}
```

```c++
void PlatformView::NotifyCreated() {
  ...
  fml::ManualResetWaitableEvent latch;
  fml::TaskRunner::RunNowOrPostTask(
      task_runners_.GetRasterTaskRunner(),
      [platform_view, &surface, &latch]() {
        surface = platform_view->CreateRenderingSurface();
        if (surface && !surface->IsValid()) {
          surface.reset();
        }
        latch.Signal();
      });
  latch.Wait();   ////------>>>> C
  ...
}
```

when the platform thread executes after A point and before B point, the raster thread preempts the CPU and obtains the right to execute, and starts thread merging (Assuming the raster task queue is empty now). When the platform thread gets the CPU again, the task continues to be put into the raster task queue, waiting for the platform thread to be processed later. But the platform thread will block at the C point.

Reply:
So if we put a mutex around RunNowOrPostTask that was shared with the thread merger, would that simplify things at all?

IOW, before A lock a mutex that MergeThreads also has to lock to make progress, and as long as you use RunNowOrPostTask (which we do) you should be ok - if you don't, you need to opt in.

Reply:
> So if we put a mutex around RunNowOrPostTask that was shared with the thread merger, would that simplify things at all?
> 
> IOW, before A lock a mutex that MergeThreads also has to lock to make progress, and as long as you use RunNowOrPostTask (which we do) you should be ok - if you don't, you need to opt in.

Yes, It's a good choice. If I understand correctly, this new function looks like the following:

```c++
void TaskRunner::RunNowOrPostTaskWithLock(fml::RefPtr<fml::TaskRunner> runner,
                                          const fml::closure& task,
                                          const std::mutex& mutex) {
  std::scoped_lock lock(mutex);
  FML_DCHECK(runner);
  if (runner->RunsTasksOnCurrentThread()) {
    task();
  } else {
    runner->PostTask(std::move(task));
  }
}
```

Reply:
The above is just the first step, we also need to flush the raster task runner before merging to make sure the sync task the platform thread post has been executed.